applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.h

```diff
@@ -0,0 +1,220 @@
+
+    /// Destructor.
+    ~AdjointNodalDisplacementResponseFunction() override;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void GetNeighboringElementPointer();
+
+
+    double CalculateValue(ModelPart& rModelPart) override;
+
+      
+    void CalculateGradient(const Element& rAdjointElem, const Matrix& rAdjointMatrix,
+                                   Vector& rResponseGradient,
+                                   ProcessInfo& rProcessInfo) override;
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+    ///@}
+ 
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    
+    void CalculateSensitivityGradient(Element& rAdjointElem,
+                                      const Variable<double>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rResponseGradient,
+                                      ProcessInfo& rProcessInfo) override;
+   
+    void CalculateSensitivityGradient(Condition& rAdjointCondition,
+                                     const Variable<double>& rVariable,
+                                     const Matrix& rDerivativesMatrix,
+                                     Vector& rResponseGradient,
+                                     ProcessInfo& rProcessInfo) override;
+
+    void CalculateSensitivityGradient(Element& rAdjointElem,
+                                      const Variable<array_1d<double,3>>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rResponseGradient,
+                                      ProcessInfo& rProcessInfo) override;
+
+    void CalculateSensitivityGradient(Condition& rAdjointCondition,
+                                      const Variable<array_1d<double,3>>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rResponseGradient,
+                                      ProcessInfo& rProcessInfo) override;
+
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+private:
+
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+    double mDisplacementValue;
+    int mIdOfTracedNode;
+    std::string mTracedDofLabel;
+    PointTypePointer  mpTracedNode;
+    Element::Pointer mpNeighboringElement;
```

Comment:
This does not seems to follow the way neighbor elements are defined in Kratos

What are you using this for?

Reply:
@philbucher I am not interested in all neighboring elements. For my purposes one element is enough. As mentioned above I need to know about one element which is bounded by the traced node in order to assemble the adjoint load into the system for the adjoint problem.